### PR TITLE
swap: Incentives swap simplification

### PR DESCRIPTION
**Pull request by ralph-pichler**
Removes the ChequeRequest message. Instead peers send cheques automatically. 
Beneficiary addresses are exchanged with a handshake in the beginning

### CODE REVIEW

File: swap/protocol.go [hunk 1]
```diff
@@ -64,9 +71,28 @@ func (s *Swap) Stop() error {
 	return nil
 }
 
+func (s *Swap) verifyHandshake(msg interface{}) error {
+	handshake := msg.(*SwapHandshakeMsg)
+	var empty common.Address
+	if handshake.Beneficiary == empty {
```
Comment (vojtechsimetka) [inline swap/protocol.go]: To me this looks like a weak check. Can handshake be not empty address but instead something invalid?

File: swap/swap.go [hunk 2]
```diff
@@ -155,22 +153,90 @@ func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
...
 	//check if balance with peer is over the payment threshold
-	if peerBalance >= s.paymentThreshold {
-		//if so, send cheque request message
-		log.Warn(fmt.Sprintf("balance for peer %s went over the payment threshold %v, requesting cheque", peer.ID().String(), s.paymentThreshold))
-		chequeRequestMessage := &ChequeRequestMsg{
-			Beneficiary: s.owner.address,
-		}
-		err = swapPeer.Send(context.TODO(), chequeRequestMessage)
-
+	if peerBalance <= -1*s.paymentThreshold {
```
Comment (vojtechsimetka) [inline swap/swap.go]: Please remove the multiplication. Just use `-s.paymentThreshold`

File: swap/swap.go [hunk 3]
```diff
@@ -155,22 +153,90 @@ func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
...
 		if err != nil {
-			log.Error(fmt.Sprintf("error while sending cheque request message to peer %s: %s", peer.ID().String(), err.Error()))
+			log.Error(fmt.Sprintf("error while sending cheque to peer %s: %s", peer.ID().String(), err.Error()))
 		}
 	}
 
 	return
 }
 
+func (s *Swap) sendCheque(peer enode.ID) error {
```
Comment (vojtechsimetka) [inline swap/swap.go:177]: This function should be behind a semaphore lock.

File: swap/swap.go [hunk 4]
```diff
@@ -155,22 +153,90 @@ func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
...
+	beneficiary := swapPeer.beneficiary
+
+	_ = s.loadCheque(peer)
+	lastCheque := s.cheques[peer]
+
+	peerBalance := s.balances[peer]
+	amount := 0 - peerBalance
+
+	// emit cheque, send to peer
+	if lastCheque == nil {
```
Comment (vojtechsimetka) [inline swap/swap.go]: I would consider extracting this into a separate function that just returns the full cheque with all the data

**Review by vojtechsimetka — CHANGES_REQUESTED** [review]

File: swap/protocol.go (hunk 1 above)
Comment (mortelli) [reply swap/protocol.go]: Added type check bool variable inspection.

File: swap/swap.go (hunk 2 above)
Comment (mortelli) [reply swap/swap.go]: Done.

File: swap/swap.go (hunk 3 above)
Comment (mortelli) [reply swap/swap.go:177]: Done.

File: swap/swap.go (hunk 4 above)
Comment (mortelli) [reply swap/swap.go]: Done.

**Review by vojtechsimetka — APPROVED** [review]

**Pull request merged**